### ISSUE TITLE
[Suggestion] Relative Maglock limitations

### ISSUE THREAD

**Issue by Xerma** [tsunamayo/Starship-EVO#2691]
Would it be possible to have the maglock limitations be relative to the entity it's on rather than having set limits?

If this isn't something you'd like to implement could we then get the limits bumped up? 5T isn't all that much if I may be honest.  And one thing others may want is to be able to have much larger ships be able to carry other 'large' ships but ships that are smaller than the host ship.

What I mean by relative limits is that a formula could be used alongside the parent entity's mass to determine how much it can carry. This may even need another value to store how much the parent is already carrying to prevent it from carrying absurd amounts of things.

Say we have 'Ship A' is the parent ship that wants to grab 'Ship B', 'Ship C', and 'Ship D':
Ship A's mass is 20,000t
Ship B's mass is 5,000t
Ship C's mass is 5,000t
Ship D's mass is 4,000t

Just for pure example purposes, here is a possible equation:
Maglock limits = 700t + (Parent entity mass * .6)
        Which then Ship A's maglock limit is 12,700t so it could carry 2 of 3 ships but not all 3.

The Core Ship from Star Wars is my prime example here:
![lyLWnVE](https://user-images.githubusercontent.com/32974671/87896577-425d8700-ca0e-11ea-9122-830aed9392b4.jpg)

Another thought is perhaps remove the limitations altogether and have the connected entity's mass contribute to the parent's mass so then you have to rely on a good amount of engines to properly move

**Comment by tsunamayo** [comment]
It is not 5T but 1000T. Also I dont want to introduce gratuitous complexity. Yes ultimately they will contribute to the mass of the two. Cheers

**Comment by ProPeach** [comment]
What are the gameplay reasons for the mass limit on the maglocks? I can understand the limits on the child entities even if they are a little frustrating, but I can't see much of a reason to limit the mass on a maglock

**Comment by tsunamayo** [comment]
Tracting a massive ship with the tiny block looks stupid, so I have to add a limit. So the bigger one also needs one, so I put a super large one so it dont gets in the way. I dont know what you guys want to do with the system, maybe you would use proper docking port instead?

**Comment by ProPeach** [comment]
So, the way I envisioned this block to work is to act as a magnet for ships to stick to other ships, stations or even asteroids. The player would pilot their ship and "land" on their desired target, then flick a switch so that the maglock sticks to the host. If the host ship moved, then the piloted ship does too and vice versa. So this block would be great to put on landing gear feet, or on the front of a "tug" ship that could move about other ships or cargo containers.

Right now, it's difficult to dock like this because of the mass limit. This limit makes no sense to me. If a smaller ship has a maglock on it, it cannot dock to a ship over 1000T (Which is quite small, I have a 150m shell that is 3600T). Which means the larger ship has to mover close to the smaller ship and pick it up which seems extremely backwards to me. I don't understand the direction of the mass limit, surely a fighter should only need a small maglock to dock itself to a carrier ship, rather than a huge maglock that depends on the size of the carrier. Personally I think the limit apply to the mass of the smallest ship.

Also, the brick maglock mass is much too low. Most fighters are 2, 3 or 5 times larger than 5T. Perhaps it should be closer to 30T? The block maglock should have no limit. 

Sometimes you would want an easy, all purpose docking solution like the maglock, but for other instances players will want a docking port like the old (pre-Kickstarter) port, where a male and female port snap together and align with the grid. I think the model for the current Maglock block is much more suited for the grid alignment docking as it has the "teeth" that can interlink with another port, the current mechanics are much more suited for a flatter shape that looks like it sticks to the flat surface of another ship/station/asteroid.

**Comment by BrickMacklin** [comment]
> Sometimes you would want an easy, all purpose docking solution like the maglock, but for other instances players will want a docking port like the old (pre-Kickstarter) port, where a male and female port snap together and align with the grid. I think the model for the current Maglock block is much more suited for the grid alignment docking as it has the "teeth" that can interlink with another port

I very much agree with this. The current lock give the impression that is how this would work and I think it is very much needed. I am finding it next to impossible to line up two ships precisely. A connector would have everything lined up in the same place every time.

**Comment by Servomotor** [comment]
I would like a docking connector/docking clamp that autopilots the small ship when it gets in range with a perfect attachment to the clamp.

Here is a mockup illustration of the idea. You see two ships here, larger one and smaller one. I would like the smaller ship to "autodock" to the larger ship when it gets within range with a perfect attachment. The idea being that we could make a small "shuttlecraft" type ship that looks as if it is part of the mother ship. Similar idea to the Star Trek Enterprise. How the Saucer Section and the Nacelle Section are actually two separate craft but usually operated being docked together. 

![20200721130145_1](https://user-images.githubusercontent.com/65745916/88091063-28c65780-cb54-11ea-83dd-54e036bd3b30.jpg)
![20200721130015_1](https://user-images.githubusercontent.com/65745916/88091067-2b28b180-cb54-11ea-946d-c3fc940a80ef.jpg)

**Comment by tsunamayo** [comment]
Yes I was also wanting it to apply to the smallest mass, I will make the change. Your other request make a lot of sense, I will see what I can do.
I still think a maximum mass make sense, but ideally with the ability to use several lock for larger build.

**Comment by ProPeach** [comment]
Awesome, thanks for replying! Which "other request" are you referring to btw? This has been a busy thread :)
I agree that if we could use multiple mag locks a limit would be much more pallateable, although the brick maglock limit is still a little too low considering it's size

**Comment by Rhedd-V** [comment]
You guys seem to be wanting to use mag locks like the landing gear in Space Engineers/Empyrion. While that would be nice sometimes, this game already has a more elegant solution to that need in the “Shipyard” part.

Set it up correctly and small ships can request docking clearance, and are then guided to a perfect landing and locked to the mothership hull where they can be repaired and be taken along for the ride.

There’s a working example on the included space stations, and it seems to work perfectly on moving vessels. It makes building and operating carriers very fun.

The mag locks seem to be intended more for moving cargo, towing smaller ships, etc. So far I’ve put them on ships that act as forklifts and tugboats, and they’re great.
